spec.bs

```diff
@@ -6,57 +6,940 @@ Status: CG-DRAFT
+
+    1. If |key| is not a [=string=] or |key|'s [=string/length=] exceeds the [=key/maximum length=], return false.
+    1. Let |environment| be |context|'s [=active window=]'s [=relevant settings object=].
+    1. [=Enqueue the following steps=] on |context|'s associated [=shared storage database/shared storage database queue=]:
+        1. Let |databaseMap| be the result of running [=obtain a shared storage bottle map=] for |environment|.
+        1. If |databaseMap| is failure, then return false.
+        1. [=map/Remove=] |databaseMap|[|key|].
+        1. Return true.
+  </div>
+
+  <div algorithm>
+    To <dfn for="shared storage database">clear all entries in the database</dfn>, given a [=/browsing context=] |context|, run the following steps:
+
+    1. Let |environment| be |context|'s [=active window=]'s [=relevant settings object=].
+    1. [=Enqueue the following steps=] on |context|'s associated [=shared storage database/shared storage database queue=]:
+        1. Let |databaseMap| be the result of running [=obtain a shared storage bottle map=] for |environment|.
+        1. If |databaseMap| is failure, then return false.
+        1. Run [=map/Clear=] on |databaseMap|.
+        1. Return true.
+  </div>
+
+  <div algorithm>
+    To <dfn for="shared storage database">retrieve all entries from the database</dfn>, given a [=/browsing context=] |context|, run the following steps:
+
+    1. Let |environment| be |context|'s [=active window=]'s [=relevant settings object=].
+    1. [=Enqueue the following steps=] on |context|'s associated [=shared storage database/shared storage database queue=]:
+        1. Let |databaseMap| be the result of running [=obtain a shared storage bottle map=] for |environment|.
+        1. If |databaseMap| is failure, then return undefined.
+        1. Let |entries| be the result of running [=map/getting the values=] on |databaseMap|.
+        1. Return |entries|.
+  </div>
+
+  <div algorithm>
+    To <dfn for="shared storage database">count entries in the database</dfn>, given a [=/browsing context=] |context|, run the following steps:
+
+    1. Let |environment| be |context|'s [=active window=]'s [=relevant settings object=].
+    1. [=Enqueue the following steps=] on |context|'s associated [=shared storage database/shared storage database queue=]:
+        1. Let |databaseMap| be the result of running [=obtain a shared storage bottle map=] for |environment|.
+        1. If |databaseMap| is failure, then return undefined.
+        1. Let |count| be |databaseMap|'s [=map/size=].
+        1. Return |count|.
+  </div>
 
-[Exposed=(Window)]
-interface SharedStorage {
-  Promise<any> set(DOMString key, DOMString value, optional SharedStorageSetMethodOptions options = {});
+  <div algorithm>
+    To <dfn for="shared storage database">purge expired entries from the database</dfn>, given a [=/browsing context=] |context|, run the following steps:
```

Comment:
What calls this algorithm?  Should we have something that says the user agent has a timer that periodically calls this algorithm?